src/gui/windows/predictor/predictor_info_dock.cpp

```diff
@@ -0,0 +1,334 @@
+    initial_state = branch_predictor->get_initial_state();
+    const machine::PredictorType predictor_type { branch_predictor->get_predictor_type() };
+    is_predictor_dynamic = predictor_type == machine::PredictorType::SMITH_1_BIT
+                            || predictor_type == machine::PredictorType::SMITH_2_BIT
+                            || predictor_type  == machine::PredictorType::SMITH_2_BIT_HYSTERESIS;
+    is_predictor_enabled = branch_predictor->get_enabled();
+
+    if (is_predictor_enabled) {
+        connect(
+            branch_predictor, &machine::BranchPredictor::total_stats_updated,
+            this, &DockPredictorInfo::update_stats);
+        connect(
+            branch_predictor, &machine::BranchPredictor::prediction_done,
+            this, &DockPredictorInfo::show_new_prediction);
+        connect(
+            core, &machine::Core::step_started,
+            this, &DockPredictorInfo::reset_colors);
+        connect(
+            branch_predictor, &machine::BranchPredictor::update_done,
+            this, &DockPredictorInfo::show_new_update);
+
+        if (is_predictor_dynamic) {
+            
+            connect(
+                branch_predictor, &machine::BranchPredictor::bhr_updated,
+                this, &DockPredictorInfo::update_bhr);
+        }
+    }
+
+    if (is_predictor_enabled) {
+        content->setDisabled(false);
+    } else {
+        content->setDisabled(true);
+    }
+
+    if (is_predictor_dynamic) {
+        label_event_predict_index_bht->setEnabled(true);
+        value_event_predict_index_bht->setEnabled(true);
+        label_event_update_index_bht->setEnabled(true);
+        value_event_update_index_bht->setEnabled(true);
+        label_bhr->setEnabled(true);
+        value_bhr->setEnabled(true);
+    } else {
+        label_event_predict_index_bht->setEnabled(false);
+        value_event_predict_index_bht->setEnabled(false);
+        label_event_update_index_bht->setEnabled(false);
+        value_event_update_index_bht->setEnabled(false);
+        label_bhr->setEnabled(false);
+        value_bhr->setEnabled(false);
+    }
+
+    if (number_of_bhr_bits == 0) {
+        label_bhr->setEnabled(false);
+        value_bhr->setEnabled(false);
+    }
+
+    clear_bhr();
+}
+
+void DockPredictorInfo::update_bhr(uint8_t number_of_bhr_bits, uint16_t register_value) {
+    if (number_of_bhr_bits > 0) {
+        QString binary_value, zero_padding;
+        binary_value = QString::number(register_value, 2);
+        zero_padding.fill('0', number_of_bhr_bits - binary_value.count());
+        value_bhr->setText("0b" + zero_padding + binary_value);
+    } else {
+        value_bhr->setText("");
+    }
+}
+
+void DockPredictorInfo::show_new_prediction(
+    uint16_t btb_index,
+    uint16_t bht_index,
+    machine::PredictionInput input,
+    machine::BranchResult result,
+    machine::BranchType branch_type) {
+    value_event_predict_instruction->setText(input.instruction.to_str());
```

Comment:
Duplication

Reply:
With the next fn

Reply:
Similar with the previous prediction / update widget duplication, the logic is the same, it just works with different widgets eg. "value_event_predict_instruction" VS "value_event_update_instruction".

The ideal solution would be probably to have the widget as a separate class and have it instantiated two times, but I feel like it would not save that much code, and the need for two separate functions to handle the two widgets would still be there.

If you'd like I can move the code into a new class, but I would prefer not to, unless there is a need to re-use it somewhere else also.

Reply:
OH, never mind. I am just blind (and tired). No need to over-engineer this. Keeps it as is.